### ISSUE TITLE
Gir submodules alternative

### ISSUE THREAD

**Issue by gkoz** [gtk-rs/gtk#311]
There is a significant drawback to having 'gir' and 'gir-files' submodules in each repo: wasted space and build time when working with several repos at once. It is desirable to find another solution which would retain ease of use (edit `Gir.toml`, `make gir`) and similar control over 'gir' and 'gir-files' revisions used to regenerate code.

**Comment by gkoz** [comment]
A possible alternative.

Use an ignored `local-gir` directory and a `gir.txt` with commit hashes. The makefile would know how to 
* `mkdir local-gir; git clone $GIR_URL local-gir/gir; git clone $GIR_FILES_URL local-gir/gir-files` if they're absent but not clobber existing ones
* check out the commits set in `gir.txt` (basically what `git submodule update` does)

You could either let `make gir` do the default thing or symlink `local-gir` somewhere and it would only check out the necessary commits if asked to (e.g. `make reset-gir`) or just build run the existing `gir` it finds.

**Comment by EPashkin** [comment]
Other alternative: optionally make `gir/target` symlink to base `gir/target`, may be ever add make target to do that.

**Comment by gkoz** [comment]
Something like this can be achieved with cargo config.

```
workdir
├── gdk
│   └── gir
├── gtk
│   └── gir
├── .cargo
│   └── config
└── target
```

Put `target-dir = "./target"` in `config` to make all crates use the same target directory. Just need to make `make` aware of where to find the binary somehow, or rewrite it to use `cargo run`.

**Comment by EPashkin** [comment]
Hard making all `target` to same directory IMHO not good idea as modules can have different version of gir.
If you think that gir better be always same then better just add new module with `gtk`, `gdk`, ..., `gir`, `gir-files` as sub-modules.

**Comment by gkoz** [comment]
You'll still have to rebuild `gir` when jumping between crates but none of its dependencies (if the lockfiles don't differ). This can be avoided by getting even more creative and using `cargo install` to make this independent of where the `target` directory is:
```make
GIR = gir/target/bin/gir
# ...
$(GIR) : gir/Cargo.toml gir/Cargo.lock gir/build.rs $(shell find gir/src -name '*.rs')
	cargo uninstall --root gir/target gir >& /dev/null || true
	cargo install --root gir/target --path gir
```

**Comment by EPashkin** [comment]
I not understand how cargo install be better.

PS. We need full rebuild `gir` on update local rustc version (every 6 week).

**Comment by gkoz** [comment]
I'll probably add a PR to make the install idea easier to understand. As to rustc upgrades the benefit of static linking by default is `gir` the binary isn't dependent on `libstd-blahblah.so`, you don't need to rebuild it just because rustc is upgraded.

**Comment by EPashkin** [comment]
Yes, it static linked, so when we try build `gir` under new `rustc` it do full rebuild as `std` version changed.

**Comment by EPashkin** [comment]
Last only true if not using `make`.
